### ISSUE TITLE
limit_req_zone with jwt embedded variables

### ISSUE THREAD

**Issue by zil1** [max-lt/nginx-jwt-module#28]
Hi there,

I'm trying to set rate limits using jwt embedded variables. But it doesn't work.
Configuration example:
```
http {
...

  limit_req_zone $jwt_claim_id zone=lreq_user:10m rate=1r/m;
  ...
  server {
    auth_jwt off;
    auth_jwt_key /etc/nginx/certs/jwt.pem file;
    ...
    location /api/ {
      auth_jwt on;
      limit_req zone=lreq_user;
      ...
    }
  }
}
```
It looks like $jwt_claim_id is defined as an empty value in the limit_req_zone directive, because with this configuration, the rate limit worked:
```
http {
...
  map $jwt_claim_id $test {
     \"100172\" 0;
     "" 1;
    }

  limit_req_zone $test zone=lreq_user:10m rate=1r/m;
  ...
  server {
    auth_jwt off;
    auth_jwt_key /etc/nginx/certs/jwt.pem file;
    ...
    location /api/ {
      auth_jwt on;
      limit_req zone=lreq_user;
      ...
    }
  }
}
```
Also strange behavior when using the auth_jwt_require directive, with this configuration nginx responds with a 403 code:
```
http {
...

  limit_req_zone $jwt_claim_id zone=lreq_user:10m rate=1r/m;
  ...
  server {
    auth_jwt off;
    auth_jwt_key /etc/nginx/certs/jwt.pem file;
    ...
    location /api/ {
      auth_jwt on;
      limit_req zone=lreq_user;
      auth_jwt_require $jwt_claim_id error=403;
      ...
    }
  }
}
```
Without the line `limit_req zone=lreq_user;` nginx works fine

**Comment by max-lt** [comment]
Hello, sorry for the late response,
I will dig into it,
my first guess would be about mdule ordering, the req limit module may be called before the jwt one so $jwt_claim_id is null when evaluating, but it's surprising.

Could you debug by sending back a header with jwt_claim_id to ensure it's set in context ? 
``` add_header X-JWT-ID $jwt_claim_id always;```

**Comment by max-lt** [comment]
After some thinking I think you almost found the solution yourself, 
the solution is to use a map like you did because `$jwt_claim_id` will only be set while reading a JWT token (so at runtime),
you need to set a map to bind `$jwt_claim_id` to a default value if not defined yet for the `limit_req_zone` directive.

```nginx
 map $jwt_claim_id $lreq_user_key {
    "" 1;
    default $jwt_claim_id;
  }
```

This should `limit_req_zone` to use `$lreq_user_key` that is 1 if `$jwt_claim_id` unset (config time) and `$jwt_claim_id` when reading a JWT that has a id property.

**Comment by zil1** [comment]
Sorry, I misunderstood how the req limit module works, and misled.
We are trying to set a limit for a specific user:
```
  map $jwt_claim_id $lreq_user_key {
      \"100172\" $jwt_claim_id;
      default "";
  }
```
I added header:
`add_header X-JWT-ID $lreq_user_key always;`

I get the correct header value, but if I add limit_req_zone for this variable, I don't get a header. 
```
http {
...
  map $jwt_claim_id $lreq_user_key {
      \"100172\" $jwt_claim_id;
      default "";
  }

  limit_req_zone $lreq_user_key zone=lreq_user:10m rate=1r/m;
  ...
  server {
    auth_jwt off;
    auth_jwt_key /etc/nginx/certs/jwt.pem file;
    ...
    location /api/ {
      auth_jwt on;
      limit_req zone=lreq_user;
      ...
    }
  }
}
```
At the same time, jwt authorization works

**Comment by max-lt** [comment]
If limit_req_zone returns no header it may be matching the default (empty string) value, try 
```
map $jwt_claim_id $lreq_user_key {
      \"100172\" $jwt_claim_id;
      default "default_zone";
  }
```

to check.

Is your jwt id (jwt_claim_id) a string or a number ?

**Comment by mechanical-rus528** [comment]
I've run into a similar problem with `limit_conn_zone`. Using something like:

```
limit_conn_zone $jwt_claim_id zone=id:10m;

server {
  limit_conn id 1;
}
```

does not actually limit the number of connections.

Using `add_header X-JWT-ID $jwt_claim_id always;` does return the expected ID, so I think your guess about module ordering might be right. Perhaps using `ngx_module_order` in `config` will help?

**Comment by mechanical-rus528** [comment]
Note that a `map` didn't do anything to help the situation. When doing

```
map $jwt_claim_id $mapped_id {
  "dummy" 0;
  default $jwt_claim_id;
}
add_header X-MAPPED-ID $mapped_id always;
```

the header is never there. So it seems like `$jwt_claim_id` is not yet set when the `map` is executed? Though I might be misunderstanding how nginx works.

**Comment by max-lt** [comment]
I can reproduce with

Config:
```
limit_req_zone $mapped_id zone=lreq_user:10m rate=1r/m;

map $jwt_claim_id $mapped_id {
  default       $jwt_claim_id;
  ""            "not-set";
}

server {
  // ...

  location = /limit-req {
    auth_jwt_key 'secret' utf8;

    add_header X-MAPPED-ID $mapped_id always;
    add_header X-JWT-ID $jwt_claim_id always;

    limit_req zone=lreq_user burst=1 nodelay;

    try_files _ =201;
  }
  // ...
}
```

JWT data:
```
{
  "sub": "1234567890",
  "name": "John Doe",
  "iat": 1516239022,
  "id": 123456
}
```

Encoded
```
export JWT=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyLCJpZCI6MTIzNDU2fQ.2CstE0GAwt2nkyOfc4rtqZJsyYzpBipOQf_dij9NGc4
```

```bash
make test-image
docker run --rm --name jwt-nginx-test -p 8000:8000 jwt-nginx-test
curl -v localhost:8000/limit-req --header "Authorization: Bearer ${JWT}"
```

Call return when the server limit_req directive is set:
```
< X-MAPPED-ID: not-set
< X-JWT-ID: 123456
```

Call return when the server limit_req directive is **not** set:
```
< X-MAPPED-ID: 123456
< X-JWT-ID: 123456
```

I will look if it's possible to choose modue order (may gives hints for #1 too).

**Comment by max-lt** [comment]
Ok I've dug it a bit, so we can register module [phases](https://nginx.org/en/docs/dev/development_guide.html#http_phases), this module is in NGX_HTTP_ACCESS_PHASE which is one phase after NGX_HTTP_PREACCESS_PHASE (phase for [ngx_http_limit_conn_module ](https://nginx.org/en/docs/http/ngx_http_limit_conn_module.html)and [ngx_http_limit_req_module](https://nginx.org/en/docs/http/ngx_http_limit_req_module.html)).

I think the best way is to execute the variable operations in one of the early phases and keep the JWT check in the access phase.

**Comment by max-lt** [comment]
It works on [this branch](https://github.com/max-lt/nginx-jwt-module/tree/28-limit_req_zone-with-jwt-embedded-variables). I'll do some more checks but it should be merged soon.

**Comment by mechanical-rus528** [comment]
Fantastic! Nice digging, and I really appreciate you looking into this. <3

**Comment by zil1** [comment]
it works, thanks!

**Comment by max-lt** [comment]
Hi guys, it's merged and published under [v3.2.2](https://github.com/max-lt/nginx-jwt-module/pkgs/container/nginx-jwt-module/98506536?tag=v3.2.2)
